assertj-core/src/main/java/org/assertj/core/api/recursive/comparison/FieldLocation.java

```diff
@@ -256,21 +266,21 @@ public boolean hasChild(FieldLocation child) {
     return child.hasParent(this);
   }
 
-  private List<String> pathsHierarchyToUseInRules() {
-    List<FieldLocation> fieldAndParentFields = list();
-    FieldLocation currentLocation = this;
-    while (!currentLocation.isRoot()) {
-      fieldAndParentFields.add(currentLocation);
-      currentLocation = currentLocation.parent();
+  private Set<String> pathsHierarchyToUseInRules() {
+    Set<String> fieldAndParentFields = newHashSet();
+    String currentPath = this.pathToUseInRules;
+    while (!isRootPath(currentPath)) {
+      fieldAndParentFields.add(currentPath);
+      currentPath = parent(currentPath);
     }
-    return fieldAndParentFields.stream()
-                               .map(fieldLocation -> fieldLocation.pathToUseInRules)
-                               .collect(toList());
+    return Collections.unmodifiableSet(fieldAndParentFields);
   }
 
-  private FieldLocation parent() {
-    List<String> parentPath = new ArrayList<>(decomposedPath);
-    parentPath.remove(decomposedPath.size() - 1);
-    return new FieldLocation(parentPath);
```

Comment:
This was allocating a lot of `ArrayList` and `FieldLocation`. I had tried `return new FieldLocation(decomposedPath.subList(0, decomposedPath.size() - 1));` but that still allocates `O(N)` `FieldLocation` as `pathsHierarchyToUseInRules` traverses the node's path to root.

Reply:
we could simplify the hash code by only using `pathToUseInRules` since the fields are more or less variations that information

Reply:
good call, `String#hashCode()` is memoized and much less expensive than also including `List<String>#hashCode()` and `Set<String>#hashCode()` which are each `O(N)`

Reply:
updated

Reply:
Going to adjust this to satisfy equals verifier error

```
-> Significant fields: equals relies on decomposedPath, but hashCode does not.
```